<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.3.450">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>The {traits.build} data standard, R package, and workflow - 29&nbsp; AusTraits tutorial</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<link href="./traits_and_climate_example.html" rel="next">
<link href="./austraits_package.html" rel="prev">
<script src="site_libs/quarto-html/quarto.js"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "sidebar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "start",
  "type": "textbox",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>


</head>

<body class="nav-sidebar floating">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
  <nav class="quarto-secondary-nav">
    <div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
      <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="./austraits_database.html">Using outputs of <code>traits.build</code></a></li><li class="breadcrumb-item"><a href="./AusTraits_tutorial.html"><span class="chapter-number">29</span>&nbsp; <span class="chapter-title">AusTraits tutorial</span></a></li></ol></nav>
      <a class="flex-grow-1" role="button" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
      </a>
      <button type="button" class="btn quarto-search-button" aria-label="" onclick="window.quartoOpenSearch();">
        <i class="bi bi-search"></i>
      </button>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal sidebar-navigation floating overflow-auto">
    <div class="pt-lg-2 mt-2 text-left sidebar-header">
    <div class="sidebar-title mb-0 py-0">
      <a href="./">The {traits.build} data standard, R package, and workflow</a> 
    </div>
      </div>
        <div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
        </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="true">
 <span class="menu-text">Introduction</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-1" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">About</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./motivation.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">Motivation</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./workflow.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">Workflow</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./usage_examples.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">Usage examples</span></span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" aria-expanded="true">
 <span class="menu-text">Data structure and standard</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-2" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./long_wide.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">5</span>&nbsp; <span class="chapter-title">Long vs Wide data</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./database_standard.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">6</span>&nbsp; <span class="chapter-title">Data standard</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./database_structure.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">7</span>&nbsp; <span class="chapter-title">Data structure</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./database_metadata.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">8</span>&nbsp; <span class="chapter-title">Overview of traits.build ontology</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./dictionary.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">9</span>&nbsp; <span class="chapter-title">Trait dictionary</span></span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" aria-expanded="true">
 <span class="menu-text">Creating with <code>traits.build</code></span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-3" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./traits_build.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">10</span>&nbsp; <span class="chapter-title">The package</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./tutorial_compilation.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">11</span>&nbsp; <span class="chapter-title">Tutorial: Example compilation</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./file_organisation.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">12</span>&nbsp; <span class="chapter-title">File organisation</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./create_dictionary.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">13</span>&nbsp; <span class="chapter-title">Creating a dictionary</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./adding_data_brief.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">14</span>&nbsp; <span class="chapter-title">Adding datasets (brief)</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./custom_R_code.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">15</span>&nbsp; <span class="chapter-title">Adding custom_R_code</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./publishing.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">16</span>&nbsp; <span class="chapter-title">Publishing</span></span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-4" aria-expanded="true">
 <span class="menu-text">Guide to adding data</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-4" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-4" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./adding_data_long.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">17</span>&nbsp; <span class="chapter-title">Adding datasets</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./data_common_issues.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">18</span>&nbsp; <span class="chapter-title">Common issues</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./tutorial_datasets.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">19</span>&nbsp; <span class="chapter-title">Tutorial: Adding datasets</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./tutorial_dataset_1.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">20</span>&nbsp; <span class="chapter-title">Tutorial 1: Adding a simple dataset</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./tutorial_dataset_2.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">21</span>&nbsp; <span class="chapter-title">Tutorial 2: Adding a more complex dataset</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./tutorial_dataset_3.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">22</span>&nbsp; <span class="chapter-title">Tutorial 3: Adding contexts and complex units</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./tutorial_dataset_4.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">23</span>&nbsp; <span class="chapter-title">Tutorial 4: Additional complexities</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./tutorial_dataset_5.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">24</span>&nbsp; <span class="chapter-title">Tutorial 5: Multiple columns for a trait</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./tutorial_dataset_6.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">25</span>&nbsp; <span class="chapter-title">Tutorial 6: Data with repeat measurements</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./tutorial_dataset_7.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">26</span>&nbsp; <span class="chapter-title">Tutorial 7: Adding long format dataset</span></span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-5" aria-expanded="true">
 <span class="menu-text">Using outputs of <code>traits.build</code></span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-5" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-5" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./austraits_database.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">27</span>&nbsp; <span class="chapter-title">AusTraits, Australia’s Plant Trait Database</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./austraits_package.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">28</span>&nbsp; <span class="chapter-title">The <code>austraits</code> package</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./AusTraits_tutorial.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text"><span class="chapter-number">29</span>&nbsp; <span class="chapter-title">AusTraits tutorial</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./traits_and_climate_example.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">30</span>&nbsp; <span class="chapter-title">Combining trait values and climate data</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./spatial_data_example.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">31</span>&nbsp; <span class="chapter-title">Tutorial: Using AusTraits with spatial data</span></span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-6" aria-expanded="true">
 <span class="menu-text">Getting help</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-6" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-6" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./help.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">32</span>&nbsp; <span class="chapter-title">Getting help</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./debugging.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">33</span>&nbsp; <span class="chapter-title">Debugging</span></span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-7" aria-expanded="true">
 <span class="menu-text">Appendices</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-7" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-7" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./csv.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">A</span>&nbsp; <span class="chapter-title">CSV files</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./yaml.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">B</span>&nbsp; <span class="chapter-title">Yaml files</span></span></a>
  </div>
</li>
      </ul>
  </li>
    </ul>
    </div>
</nav>
<div id="quarto-sidebar-glass" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">Table of contents</h2>
   
  <ul>
  <li><a href="#introduction" id="toc-introduction" class="nav-link active" data-scroll-target="#introduction"><span class="header-section-number">29.1</span> Introduction</a></li>
  <li><a href="#download-austraits-data" id="toc-download-austraits-data" class="nav-link" data-scroll-target="#download-austraits-data"><span class="header-section-number">29.2</span> Download AusTraits data</a></li>
  <li><a href="#a-first-look-at-data" id="toc-a-first-look-at-data" class="nav-link" data-scroll-target="#a-first-look-at-data"><span class="header-section-number">29.3</span> A first look at data</a></li>
  <li><a href="#database_structure" id="toc-database_structure" class="nav-link" data-scroll-target="#database_structure"><span class="header-section-number">29.4</span> The database structure</a>
  <ul class="collapse">
  <li><a href="#traits-table" id="toc-traits-table" class="nav-link" data-scroll-target="#traits-table">Traits table</a></li>
  <li><a href="#ancillary-data-tables" id="toc-ancillary-data-tables" class="nav-link" data-scroll-target="#ancillary-data-tables">Ancillary data tables</a></li>
  </ul></li>
  <li><a href="#exploring-austraits" id="toc-exploring-austraits" class="nav-link" data-scroll-target="#exploring-austraits"><span class="header-section-number">29.5</span> Exploring AusTraits</a>
  <ul class="collapse">
  <li><a href="#austraits.r-function-reference" id="toc-austraits.r-function-reference" class="nav-link" data-scroll-target="#austraits.r-function-reference">austraits.R function reference</a></li>
  <li><a href="#breadth-and-depth-of-austraits-data" id="toc-breadth-and-depth-of-austraits-data" class="nav-link" data-scroll-target="#breadth-and-depth-of-austraits-data">breadth and depth of AusTraits data</a></li>
  <li><a href="#filtering-data" id="toc-filtering-data" class="nav-link" data-scroll-target="#filtering-data">Filtering data</a></li>
  </ul></li>
  <li><a href="#wrangling-data-joining-tables" id="toc-wrangling-data-joining-tables" class="nav-link" data-scroll-target="#wrangling-data-joining-tables"><span class="header-section-number">29.6</span> Wrangling data: joining tables</a>
  <ul class="collapse">
  <li><a href="#locations" id="toc-locations" class="nav-link" data-scroll-target="#locations">Locations</a></li>
  <li><a href="#contexts" id="toc-contexts" class="nav-link" data-scroll-target="#contexts">Contexts</a></li>
  <li><a href="#methods" id="toc-methods" class="nav-link" data-scroll-target="#methods">Methods</a></li>
  <li><a href="#taxa" id="toc-taxa" class="nav-link" data-scroll-target="#taxa">Taxa</a></li>
  </ul></li>
  <li><a href="#plotting-data" id="toc-plotting-data" class="nav-link" data-scroll-target="#plotting-data"><span class="header-section-number">29.7</span> Plotting data</a>
  <ul class="collapse">
  <li><a href="#plotting-locations" id="toc-plotting-locations" class="nav-link" data-scroll-target="#plotting-locations">Plotting locations</a></li>
  <li><a href="#an-example-looking-at-trait-climate-gradients" id="toc-an-example-looking-at-trait-climate-gradients" class="nav-link" data-scroll-target="#an-example-looking-at-trait-climate-gradients">An example looking at trait-climate gradients</a></li>
  <li><a href="#an-example-with-ala-data" id="toc-an-example-with-ala-data" class="nav-link" data-scroll-target="#an-example-with-ala-data">An example with ALA data</a></li>
  </ul></li>
  <li><a href="#summarising-data" id="toc-summarising-data" class="nav-link" data-scroll-target="#summarising-data"><span class="header-section-number">29.8</span> Summarising data</a>
  <ul class="collapse">
  <li><a href="#summarising-traits-by-taxon-counts-dataset_id-etc." id="toc-summarising-traits-by-taxon-counts-dataset_id-etc." class="nav-link" data-scroll-target="#summarising-traits-by-taxon-counts-dataset_id-etc.">Summarising traits by taxon counts, dataset_id, etc.</a></li>
  </ul></li>
  <li><a href="#a-complexity-pivoting-datasets" id="toc-a-complexity-pivoting-datasets" class="nav-link" data-scroll-target="#a-complexity-pivoting-datasets"><span class="header-section-number">29.9</span> A complexity: pivoting datasets</a></li>
  <li><a href="#intepreting-trait-names-taxon-names" id="toc-intepreting-trait-names-taxon-names" class="nav-link" data-scroll-target="#intepreting-trait-names-taxon-names"><span class="header-section-number">29.10</span> Intepreting trait names, taxon names</a>
  <ul class="collapse">
  <li><a href="#trait-dictionary" id="toc-trait-dictionary" class="nav-link" data-scroll-target="#trait-dictionary">Trait dictionary</a></li>
  <li><a href="#understanding-taxon-names" id="toc-understanding-taxon-names" class="nav-link" data-scroll-target="#understanding-taxon-names">Understanding taxon names</a></li>
  </ul></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title"><span class="chapter-number">29</span>&nbsp; <span class="chapter-title">AusTraits tutorial</span></h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  

</header>

<section id="introduction" class="level2" data-number="29.1">
<h2 data-number="29.1" class="anchored" data-anchor-id="introduction"><span class="header-section-number">29.1</span> Introduction</h2>
<p>With more than 1.8 million data records, AusTraits is Australia’s <a href="austraits_database.html#plant_database">largest plant trait database</a>, created using the [<code>traits.build</code>] package(https://github.com/traitecoevo/traits.build)</p>
<p>The following tutorial introduces:</p>
<ul>
<li><a href="#database_structure">the database structure</a><br>
</li>
<li></li>
<li>additional examples of analyses using the database<br>
</li>
<li></li>
</ul>
<p>To access the Github repositories for individual components, see:</p>
<ul>
<li><p>the database structure, <a href="https://github.com/traitecoevo/traits.build"><code>traits.build</code></a></p></li>
<li><p>the database contents, <a href="https://github.com/traitecoevo/austraits.build"><code>austraits.build</code></a></p></li>
<li><p>an R package for exploring and wrangling the data, <a href="https://github.com/traitecoevo/austraits"><code>austraits</code></a></p></li>
</ul>
</section>
<section id="download-austraits-data" class="level2" data-number="29.2">
<h2 data-number="29.2" class="anchored" data-anchor-id="download-austraits-data"><span class="header-section-number">29.2</span> Download AusTraits data</h2>
<p>Before you begin, download (or build) the latest AusTraits data, by one of the methods described <a href="austraits_database.html#access_data">here</a></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(readr)</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyr)</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(dplyr)</span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(stringr)</span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(austraits)</span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(APCalign)</span>
<span id="cb1-7"><a href="#cb1-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-8"><a href="#cb1-8" aria-hidden="true" tabindex="-1"></a><span class="fu">source</span>(<span class="st">"data/extra_functions.R"</span>)</span>
<span id="cb1-9"><a href="#cb1-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-10"><a href="#cb1-10" aria-hidden="true" tabindex="-1"></a>austraits <span class="ot">&lt;-</span> <span class="fu">read_rds</span>(<span class="st">"data/austraits_5.Rds"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="a-first-look-at-data" class="level2" data-number="29.3">
<h2 data-number="29.3" class="anchored" data-anchor-id="a-first-look-at-data"><span class="header-section-number">29.3</span> A first look at data</h2>
<p>Some traits have a vast amount of data:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a>austraits<span class="sc">$</span>traits <span class="sc">%&gt;%</span> </span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">filter</span>(trait_name <span class="sc">==</span> <span class="st">"leaf_N_per_dry_mass"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">distinct</span>(taxon_name) <span class="sc">%&gt;%</span> <span class="fu">nrow</span>()</span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-5"><a href="#cb2-5" aria-hidden="true" tabindex="-1"></a>[<span class="dv">1</span>] <span class="dv">2262</span></span>
<span id="cb2-6"><a href="#cb2-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-7"><a href="#cb2-7" aria-hidden="true" tabindex="-1"></a>austraits<span class="sc">::</span><span class="fu">plot_trait_distribution_beeswarm</span>(austraits, <span class="st">"leaf_N_per_dry_mass"</span>, <span class="at">y_axis_category =</span> <span class="st">"dataset_id"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Others much less:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a>austraits<span class="sc">::</span><span class="fu">lookup_trait</span>(austraits, <span class="st">"hydraulic"</span>)</span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a> [<span class="dv">1</span>] <span class="st">"sapwood_specific_hydraulic_conductivity_theoretical"</span> <span class="st">"stem_vessel_diameter_hydraulic"</span>                     </span>
<span id="cb3-4"><a href="#cb3-4" aria-hidden="true" tabindex="-1"></a> [<span class="dv">3</span>] <span class="st">"leaf_hydraulic_vulnerability"</span>                        <span class="st">"leaf_specific_hydraulic_conductance"</span>                </span>
<span id="cb3-5"><a href="#cb3-5" aria-hidden="true" tabindex="-1"></a> [<span class="dv">5</span>] <span class="st">"leaf_specific_hydraulic_conductivity"</span>                <span class="st">"sapwood_specific_hydraulic_conductivity"</span>            </span>
<span id="cb3-6"><a href="#cb3-6" aria-hidden="true" tabindex="-1"></a> [<span class="dv">7</span>] <span class="st">"stem_specific_hydraulic_conductivity"</span>                <span class="st">"stem_hydraulic_conductivity"</span>                        </span>
<span id="cb3-7"><a href="#cb3-7" aria-hidden="true" tabindex="-1"></a> [<span class="dv">9</span>] <span class="st">"hydraulic_safety_margin_50"</span>                          <span class="st">"hydraulic_safety_margin_88"</span> </span>
<span id="cb3-8"><a href="#cb3-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-9"><a href="#cb3-9" aria-hidden="true" tabindex="-1"></a>austraits<span class="sc">$</span>traits <span class="sc">%&gt;%</span> </span>
<span id="cb3-10"><a href="#cb3-10" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">filter</span>(trait_name <span class="sc">==</span> <span class="st">"sapwood_specific_hydraulic_conductivity"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb3-11"><a href="#cb3-11" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">distinct</span>(taxon_name) <span class="sc">%&gt;%</span> <span class="fu">nrow</span>()</span>
<span id="cb3-12"><a href="#cb3-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-13"><a href="#cb3-13" aria-hidden="true" tabindex="-1"></a>[<span class="dv">1</span>] <span class="dv">179</span></span>
<span id="cb3-14"><a href="#cb3-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-15"><a href="#cb3-15" aria-hidden="true" tabindex="-1"></a>austraits<span class="sc">::</span><span class="fu">plot_trait_distribution_beeswarm</span>(austraits, <span class="st">"sapwood_specific_hydraulic_conductivity"</span>, <span class="at">y_axis_category =</span> <span class="st">"dataset_id"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Most field-collected data are linked to site coordinates:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a>data <span class="ot">&lt;-</span></span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a>  austraits <span class="sc">%&gt;%</span></span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true" tabindex="-1"></a>    austraits<span class="sc">::</span><span class="fu">extract_trait</span>(<span class="at">trait_names =</span> <span class="st">"sapwood_specific_hydraulic_conductivity"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb4-4"><a href="#cb4-4" aria-hidden="true" tabindex="-1"></a>    austraits<span class="sc">::</span><span class="fu">join_locations</span>()</span>
<span id="cb4-5"><a href="#cb4-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-6"><a href="#cb4-6" aria-hidden="true" tabindex="-1"></a>data<span class="sc">$</span>traits <span class="sc">%&gt;%</span> austraits<span class="sc">::</span><span class="fu">plot_locations</span>(<span class="st">"trait_name"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Or:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a>data <span class="ot">&lt;-</span></span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a>  austraits <span class="sc">%&gt;%</span></span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true" tabindex="-1"></a>     austraits<span class="sc">::</span><span class="fu">extract_taxa</span>(<span class="at">taxon_name =</span> <span class="st">"Acacia aneura"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb5-4"><a href="#cb5-4" aria-hidden="true" tabindex="-1"></a>     austraits<span class="sc">::</span><span class="fu">join_locations</span>()</span>
<span id="cb5-5"><a href="#cb5-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-6"><a href="#cb5-6" aria-hidden="true" tabindex="-1"></a>data<span class="sc">$</span>traits <span class="sc">%&gt;%</span> austraits<span class="sc">::</span><span class="fu">plot_locations</span>(<span class="st">"taxon_name"</span>) <span class="co"># 1 taxon, so 1 plot</span></span>
<span id="cb5-7"><a href="#cb5-7" aria-hidden="true" tabindex="-1"></a>data<span class="sc">$</span>traits <span class="sc">%&gt;%</span> austraits<span class="sc">::</span><span class="fu">plot_locations</span>(<span class="st">"dataset_id"</span>) <span class="co"># 1 plot for each dataset_id</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>General patterns of taxon distribution are immediately apparent:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a>data <span class="ot">&lt;-</span></span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a>  austraits <span class="sc">%&gt;%</span></span>
<span id="cb6-3"><a href="#cb6-3" aria-hidden="true" tabindex="-1"></a>     austraits<span class="sc">::</span><span class="fu">extract_taxa</span>(<span class="at">genus =</span> <span class="st">"Hibbertia"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb6-4"><a href="#cb6-4" aria-hidden="true" tabindex="-1"></a>     austraits<span class="sc">::</span><span class="fu">join_locations</span>() <span class="sc">%&gt;%</span></span>
<span id="cb6-5"><a href="#cb6-5" aria-hidden="true" tabindex="-1"></a>     austraits<span class="sc">::</span><span class="fu">join_taxonomy</span>()</span>
<span id="cb6-6"><a href="#cb6-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-7"><a href="#cb6-7" aria-hidden="true" tabindex="-1"></a>data<span class="sc">$</span>traits <span class="sc">%&gt;%</span> austraits<span class="sc">::</span><span class="fu">plot_locations</span>(<span class="st">"genus"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Or to see how much data is available:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a>austraits<span class="sc">::</span><span class="fu">lookup_trait</span>(austraits, <span class="st">"_N_"</span>) <span class="ot">-&gt;</span> N_traits</span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-3"><a href="#cb7-3" aria-hidden="true" tabindex="-1"></a>austraits<span class="sc">$</span>traits <span class="sc">%&gt;%</span> </span>
<span id="cb7-4"><a href="#cb7-4" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">filter</span>(trait_name <span class="sc">%in%</span> N_traits) <span class="sc">%&gt;%</span> </span>
<span id="cb7-5"><a href="#cb7-5" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">distinct</span>(taxon_name, trait_name) <span class="sc">%&gt;%</span> </span>
<span id="cb7-6"><a href="#cb7-6" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">mutate</span>(<span class="at">counter =</span> <span class="dv">1</span>) <span class="sc">%&gt;%</span> </span>
<span id="cb7-7"><a href="#cb7-7" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">group_by</span>(trait_name) <span class="sc">%&gt;%</span> </span>
<span id="cb7-8"><a href="#cb7-8" aria-hidden="true" tabindex="-1"></a>    dplyr<span class="sc">::</span><span class="fu">mutate</span>(<span class="at">counter =</span> <span class="fu">sum</span>(counter)) <span class="sc">%&gt;%</span> </span>
<span id="cb7-9"><a href="#cb7-9" aria-hidden="true" tabindex="-1"></a>    dplyr<span class="sc">::</span><span class="fu">distinct</span>(trait_name, counter) <span class="sc">%&gt;%</span> </span>
<span id="cb7-10"><a href="#cb7-10" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">ungroup</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="database_structure" class="level2" data-number="29.4">
<h2 data-number="29.4" class="anchored" data-anchor-id="database_structure"><span class="header-section-number">29.4</span> The database structure</h2>
<p>The {traits.build} R package is the workflow that builds AusTraits from its component datasets.</p>
<p>The database is output as a collection of relational tables, as described <a href="database_structure.html">here</a>.</p>
<section id="traits-table" class="level3">
<h3 class="anchored" data-anchor-id="traits-table">Traits table</h3>
<p>The core AusTraits table is the traits table. It is in “long” format, with each row documenting a single trait measurement.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a>austraits<span class="sc">$</span>traits</span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true" tabindex="-1"></a><span class="co"># A tibble: 1,253,250 × 24</span></span>
<span id="cb8-4"><a href="#cb8-4" aria-hidden="true" tabindex="-1"></a>   dataset_id taxon_name      observation_id trait_name value unit  entity_type value_type basis_of_value replicates basis_of_record life_stage population_id individual_id temporal_id source_id location_id entity_context_id</span>
<span id="cb8-5"><a href="#cb8-5" aria-hidden="true" tabindex="-1"></a>   <span class="sc">&lt;</span>chr<span class="sc">&gt;</span>      <span class="er">&lt;</span>chr<span class="sc">&gt;</span>           <span class="er">&lt;</span>chr<span class="sc">&gt;</span>          <span class="er">&lt;</span>chr<span class="sc">&gt;</span>      <span class="er">&lt;</span>chr<span class="sc">&gt;</span> <span class="er">&lt;</span>chr<span class="sc">&gt;</span> <span class="er">&lt;</span>chr<span class="sc">&gt;</span>       <span class="er">&lt;</span>chr<span class="sc">&gt;</span>      <span class="er">&lt;</span>chr<span class="sc">&gt;</span>          <span class="er">&lt;</span>chr<span class="sc">&gt;</span>      <span class="er">&lt;</span>chr<span class="sc">&gt;</span>           <span class="er">&lt;</span>chr<span class="sc">&gt;</span>      <span class="er">&lt;</span>chr<span class="sc">&gt;</span>         <span class="er">&lt;</span>chr<span class="sc">&gt;</span>         <span class="er">&lt;</span>chr<span class="sc">&gt;</span>       <span class="er">&lt;</span>chr<span class="sc">&gt;</span>     <span class="er">&lt;</span>chr<span class="sc">&gt;</span>       <span class="er">&lt;</span>chr<span class="sc">&gt;</span>            </span>
<span id="cb8-6"><a href="#cb8-6" aria-hidden="true" tabindex="-1"></a> <span class="dv">1</span> ABRS_1981  Acanthocarpus … <span class="dv">0001</span>           leaf_comp… simp… <span class="cn">NA</span>    species     mode       expert_score   <span class="cn">NA</span>         preserved_spec… adult      <span class="cn">NA</span>            <span class="cn">NA</span>            <span class="cn">NA</span>          <span class="cn">NA</span>        <span class="cn">NA</span>          <span class="cn">NA</span>               </span>
<span id="cb8-7"><a href="#cb8-7" aria-hidden="true" tabindex="-1"></a> <span class="dv">2</span> ABRS_1981  Acanthocarpus … <span class="dv">0001</span>           leaf_leng… <span class="dv">3</span>     mm    species     minimum    measurement    <span class="cn">NA</span>         preserved_spec… adult      <span class="cn">NA</span>            <span class="cn">NA</span>            <span class="cn">NA</span>          <span class="cn">NA</span>        <span class="cn">NA</span>          <span class="cn">NA</span>               </span>
<span id="cb8-8"><a href="#cb8-8" aria-hidden="true" tabindex="-1"></a> <span class="dv">3</span> ABRS_1981  Acanthocarpus … <span class="dv">0001</span>           leaf_leng… <span class="dv">15</span>    mm    species     maximum    measurement    <span class="cn">NA</span>         preserved_spec… adult      <span class="cn">NA</span>            <span class="cn">NA</span>            <span class="cn">NA</span>          <span class="cn">NA</span>        <span class="cn">NA</span>          <span class="cn">NA</span>               </span>
<span id="cb8-9"><a href="#cb8-9" aria-hidden="true" tabindex="-1"></a> <span class="dv">4</span> ABRS_1981  Acanthocarpus … <span class="dv">0001</span>           seed_heig… <span class="dv">3</span>     mm    species     maximum    measurement    <span class="cn">NA</span>         preserved_spec… adult      <span class="cn">NA</span>            <span class="cn">NA</span>            <span class="cn">NA</span>          <span class="cn">NA</span>        <span class="cn">NA</span>          <span class="cn">NA</span>               </span>
<span id="cb8-10"><a href="#cb8-10" aria-hidden="true" tabindex="-1"></a> <span class="dv">5</span> ABRS_1981  Acanthocarpus … <span class="dv">0001</span>           seed_leng… <span class="dv">3</span>     mm    species     maximum    measurement    <span class="cn">NA</span>         preserved_spec… adult      <span class="cn">NA</span>            <span class="cn">NA</span>            <span class="cn">NA</span>          <span class="cn">NA</span>        <span class="cn">NA</span>          <span class="cn">NA</span>               </span>
<span id="cb8-11"><a href="#cb8-11" aria-hidden="true" tabindex="-1"></a> <span class="dv">6</span> ABRS_1981  Acanthocarpus … <span class="dv">0001</span>           seed_width <span class="dv">3</span>     mm    species     maximum    measurement    <span class="cn">NA</span>         preserved_spec… adult      <span class="cn">NA</span>            <span class="cn">NA</span>            <span class="cn">NA</span>          <span class="cn">NA</span>        <span class="cn">NA</span>          <span class="cn">NA</span>               </span>
<span id="cb8-12"><a href="#cb8-12" aria-hidden="true" tabindex="-1"></a> <span class="dv">7</span> ABRS_1981  Acanthocarpus … <span class="dv">0002</span>           leaf_comp… simp… <span class="cn">NA</span>    species     mode       expert_score   <span class="cn">NA</span>         preserved_spec… adult      <span class="cn">NA</span>            <span class="cn">NA</span>            <span class="cn">NA</span>          <span class="cn">NA</span>        <span class="cn">NA</span>          <span class="cn">NA</span>               </span>
<span id="cb8-13"><a href="#cb8-13" aria-hidden="true" tabindex="-1"></a> <span class="dv">8</span> ABRS_1981  Acanthocarpus … <span class="dv">0002</span>           leaf_leng… <span class="dv">4</span>     mm    species     minimum    measurement    <span class="cn">NA</span>         preserved_spec… adult      <span class="cn">NA</span>            <span class="cn">NA</span>            <span class="cn">NA</span>          <span class="cn">NA</span>        <span class="cn">NA</span>          <span class="cn">NA</span>               </span>
<span id="cb8-14"><a href="#cb8-14" aria-hidden="true" tabindex="-1"></a> <span class="dv">9</span> ABRS_1981  Acanthocarpus … <span class="dv">0002</span>           leaf_leng… <span class="dv">12</span>    mm    species     maximum    measurement    <span class="cn">NA</span>         preserved_spec… adult      <span class="cn">NA</span>            <span class="cn">NA</span>            <span class="cn">NA</span>          <span class="cn">NA</span>        <span class="cn">NA</span>          <span class="cn">NA</span>               </span>
<span id="cb8-15"><a href="#cb8-15" aria-hidden="true" tabindex="-1"></a><span class="dv">10</span> ABRS_1981  Acanthocarpus … <span class="dv">0002</span>           seed_leng… <span class="dv">4</span>     mm    species     maximum    measurement    <span class="cn">NA</span>         preserved_spec… adult      <span class="cn">NA</span>            <span class="cn">NA</span>            <span class="cn">NA</span>          <span class="cn">NA</span>        <span class="cn">NA</span>          <span class="cn">NA</span>               </span>
<span id="cb8-16"><a href="#cb8-16" aria-hidden="true" tabindex="-1"></a><span class="co"># ℹ 1,253,240 more rows</span></span>
<span id="cb8-17"><a href="#cb8-17" aria-hidden="true" tabindex="-1"></a><span class="co"># ℹ 6 more variables: plot_id &lt;chr&gt;, treatment_id &lt;chr&gt;, collection_date &lt;chr&gt;, measurement_remarks &lt;chr&gt;, method_id &lt;chr&gt;, original_name &lt;chr&gt;</span></span>
<span id="cb8-18"><a href="#cb8-18" aria-hidden="true" tabindex="-1"></a><span class="co"># ℹ Use `print(n = ...)` to see more rows</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The columns include:</p>
<ul>
<li>core columns&nbsp;
<ul>
<li>dataset_id&nbsp;</li>
<li>taxon_name&nbsp;</li>
<li>trait_name&nbsp;</li>
<li>value (trait value)&nbsp;</li>
</ul></li>
<li>entity &amp; value metadata</li>
<li>identifiers that provide links to ancillary tables with additional metadata</li>
</ul>
</section>
<section id="ancillary-data-tables" class="level3">
<h3 class="anchored" data-anchor-id="ancillary-data-tables">Ancillary data tables</h3>
<p>The remaining metadata accompanying each trait record is recorded across multiple relational tables.</p>
<p>These include:&nbsp;</p>
<ul>
<li>austraits$locations&nbsp;</li>
<li>austraits$contexts&nbsp;</li>
<li>austraits$methods&nbsp;</li>
<li>austraits$taxa&nbsp;</li>
<li>austraits$taxonomic_updates&nbsp;</li>
<li>austraits$contributors&nbsp;</li>
</ul>
<p>Like the core <code>traits</code> table, each is in ‘long’ format.</p>
<p>The tables <code>locations</code>, <code>contexts</code>, <code>methods</code> and <code>taxa</code> include metadata that links seamlessly to individual rows within <code>traits</code>.</p>
<table class="table">
<colgroup>
<col style="width: 19%">
<col style="width: 38%">
<col style="width: 42%">
</colgroup>
<thead>
<tr class="header">
<th>table</th>
<th>metadata in table</th>
<th>columns that link to austraits$traits</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>locations</td>
<td>location name, location properties, latitude, longitude</td>
<td>dataset_id, location_id</td>
</tr>
<tr class="even">
<td>contexts</td>
<td>context name, context category (method context, temporal, entity context, plot, treatment), context property</td>
<td>dataset_id, link_id (identifier to link to: method_context_id, temporal_id, entity_context_id, plot_id, treatment_id), link_vals (identifier value to link to)</td>
</tr>
<tr class="odd">
<td>methods</td>
<td>dataset description, dataset sampling strategy, trait collection method, data collectors, data curators, dataset citation, source_id &amp; citation</td>
<td>dataset_id, trait_name, method_id</td>
</tr>
<tr class="even">
<td>taxa</td>
<td>genus, family, scientific name, APC/APNI taxon concept/taxon name identifiers</td>
<td></td>
</tr>
</tbody>
</table>
</section>
</section>
<section id="exploring-austraits" class="level2" data-number="29.5">
<h2 data-number="29.5" class="anchored" data-anchor-id="exploring-austraits"><span class="header-section-number">29.5</span> Exploring AusTraits</h2>
<p>With 1,800,000+ rows of trait values in the main traits table, knowing how to explore the contents is essential.</p>
<p>The R package <a href="https://github.com/traitecoevo/austraits">austraits</a> offers a collection of functions to explore and wrangle AusTraits data - or indeed any data using the traits.build format.</p>
<section id="austraits.r-function-reference" class="level3">
<h3 class="anchored" data-anchor-id="austraits.r-function-reference">austraits.R function reference</h3>
<p>Reference guide to: <a href="https://traitecoevo.github.io/austraits/reference/index.html">austraits functions</a></p>
<p>(There is also a standalone <a href="austraits_package.html">austraits introduction</a>, which repeats some of the same material covered here.)</p>
</section>
<section id="breadth-and-depth-of-austraits-data" class="level3">
<h3 class="anchored" data-anchor-id="breadth-and-depth-of-austraits-data">breadth and depth of AusTraits data</h3>
<p>Use the function <code>summarise_austraits</code> to output summaries of total records, datasets with records, and taxa with records across <code>families</code>, <code>genera</code> or <code>traits</code>:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a>austraits<span class="sc">::</span><span class="fu">summarise_austraits</span>(austraits, <span class="st">"family"</span>)</span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a>austraits<span class="sc">::</span><span class="fu">summarise_austraits</span>(austraits, <span class="st">"genus"</span>)</span>
<span id="cb9-3"><a href="#cb9-3" aria-hidden="true" tabindex="-1"></a>austraits<span class="sc">::</span><span class="fu">summarise_austraits</span>(austraits, <span class="st">"trait_name"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Alternatively you can look up traits that contain a specific search term:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a>austraits<span class="sc">::</span><span class="fu">lookup_trait</span>(austraits, <span class="st">"leaf"</span>)</span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a>austraits<span class="sc">::</span><span class="fu">lookup_trait</span>(austraits, <span class="st">"_N_"</span>) <span class="co"># elemental contents use their symbol and are always in the middle of a trait name</span></span>
<span id="cb10-3"><a href="#cb10-3" aria-hidden="true" tabindex="-1"></a>austraits<span class="sc">::</span><span class="fu">lookup_trait</span>(austraits, <span class="st">"photo"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="filtering-data" class="level3">
<h3 class="anchored" data-anchor-id="filtering-data">Filtering data</h3>
<p>There are a trio of austraits.R functions to filter data by <code>trait_name</code>, <code>taxon_name</code> or <code>dataset_id</code>:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a>leaf_mass_per_area_values <span class="ot">&lt;-</span></span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true" tabindex="-1"></a>  austraits <span class="sc">%&gt;%</span> </span>
<span id="cb11-3"><a href="#cb11-3" aria-hidden="true" tabindex="-1"></a>     austraits<span class="sc">::</span><span class="fu">extract_trait</span>(<span class="at">trait_names =</span> <span class="fu">c</span>(<span class="st">"leaf_mass_per_area"</span>))</span>
<span id="cb11-4"><a href="#cb11-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-5"><a href="#cb11-5" aria-hidden="true" tabindex="-1"></a>Westoby_2014_values <span class="ot">&lt;-</span></span>
<span id="cb11-6"><a href="#cb11-6" aria-hidden="true" tabindex="-1"></a>  austraits <span class="sc">%&gt;%</span></span>
<span id="cb11-7"><a href="#cb11-7" aria-hidden="true" tabindex="-1"></a>     austraits<span class="sc">::</span><span class="fu">extract_dataset</span>(<span class="st">"Westoby_2014"</span>)</span>
<span id="cb11-8"><a href="#cb11-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-9"><a href="#cb11-9" aria-hidden="true" tabindex="-1"></a>Eucalyptus_values <span class="ot">&lt;-</span></span>
<span id="cb11-10"><a href="#cb11-10" aria-hidden="true" tabindex="-1"></a>  austraits <span class="sc">%&gt;%</span></span>
<span id="cb11-11"><a href="#cb11-11" aria-hidden="true" tabindex="-1"></a>     austraits<span class="sc">::</span><span class="fu">extract_taxa</span>(<span class="at">genus =</span> <span class="st">"Eucalyptus"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>You can alternatively simply filter by whatever combination of values you are searching for:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a>my_data <span class="ot">&lt;-</span></span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true" tabindex="-1"></a>  austraits<span class="sc">$</span>traits <span class="sc">%&gt;%</span></span>
<span id="cb12-3"><a href="#cb12-3" aria-hidden="true" tabindex="-1"></a>    dplyr<span class="sc">::</span><span class="fu">filter</span>(stringr<span class="sc">::</span><span class="fu">str_detect</span>(taxon_name, <span class="st">"Eucalyptus"</span>)) <span class="sc">%&gt;%</span></span>
<span id="cb12-4"><a href="#cb12-4" aria-hidden="true" tabindex="-1"></a>    dplyr<span class="sc">::</span><span class="fu">filter</span>(trait_name <span class="sc">%in%</span> <span class="fu">c</span>(<span class="st">"leaf_mass_per_area"</span>, <span class="st">"leaf_length"</span>, <span class="st">"leaf_thickness"</span>)) <span class="sc">%&gt;%</span></span>
<span id="cb12-5"><a href="#cb12-5" aria-hidden="true" tabindex="-1"></a>    dplyr<span class="sc">::</span><span class="fu">filter</span>(basis_of_record <span class="sc">==</span> <span class="st">"field"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
</section>
<section id="wrangling-data-joining-tables" class="level2" data-number="29.6">
<h2 data-number="29.6" class="anchored" data-anchor-id="wrangling-data-joining-tables"><span class="header-section-number">29.6</span> Wrangling data: joining tables</h2>
<section id="locations" class="level3">
<h3 class="anchored" data-anchor-id="locations">Locations</h3>
<p>To just add location name, latitude, and longitude:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a>traits_with_lat_long <span class="ot">&lt;-</span> </span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a>  austraits <span class="sc">%&gt;%</span> austraits<span class="sc">::</span><span class="fu">join_locations</span>()</span>
<span id="cb13-3"><a href="#cb13-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-4"><a href="#cb13-4" aria-hidden="true" tabindex="-1"></a>traits_with_lat_long<span class="sc">$</span>traits</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>To add all location properties:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb14"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a>locations <span class="ot">&lt;-</span> </span>
<span id="cb14-2"><a href="#cb14-2" aria-hidden="true" tabindex="-1"></a>  austraits<span class="sc">$</span>locations <span class="sc">%&gt;%</span></span>
<span id="cb14-3"><a href="#cb14-3" aria-hidden="true" tabindex="-1"></a>  tidyr<span class="sc">::</span><span class="fu">pivot_wider</span>(<span class="at">names_from =</span> location_property, <span class="at">values_from =</span> value)</span>
<span id="cb14-4"><a href="#cb14-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb14-5"><a href="#cb14-5" aria-hidden="true" tabindex="-1"></a>traits_with_locations <span class="ot">&lt;-</span> </span>
<span id="cb14-6"><a href="#cb14-6" aria-hidden="true" tabindex="-1"></a>  austraits<span class="sc">$</span>traits <span class="sc">%&gt;%</span> </span>
<span id="cb14-7"><a href="#cb14-7" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">left_join</span>(locations, <span class="at">by =</span> <span class="fu">c</span>(<span class="st">"dataset_id"</span>, <span class="st">"location_id"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The number of location properties is of course long. You’ll possible want to check which location properties you care about and just merge those:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a>locations_properties <span class="ot">&lt;-</span> </span>
<span id="cb15-2"><a href="#cb15-2" aria-hidden="true" tabindex="-1"></a>  austraits<span class="sc">$</span>locations <span class="sc">%&gt;%</span> </span>
<span id="cb15-3"><a href="#cb15-3" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">distinct</span>(location_property)</span>
<span id="cb15-4"><a href="#cb15-4" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb15-5"><a href="#cb15-5" aria-hidden="true" tabindex="-1"></a>locations <span class="ot">&lt;-</span> </span>
<span id="cb15-6"><a href="#cb15-6" aria-hidden="true" tabindex="-1"></a>  austraits<span class="sc">$</span>locations <span class="sc">%&gt;%</span></span>
<span id="cb15-7"><a href="#cb15-7" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">filter</span>(location_property <span class="sc">%in%</span> <span class="fu">c</span>(<span class="st">"latitude (deg)"</span>, <span class="st">"longitude (deg)"</span>, <span class="st">"description"</span>, <span class="st">"aridity index (MAP/PET)"</span>, <span class="st">"soil type"</span>, <span class="st">"fire history"</span>)) <span class="sc">%&gt;%</span></span>
<span id="cb15-8"><a href="#cb15-8" aria-hidden="true" tabindex="-1"></a>  tidyr<span class="sc">::</span><span class="fu">pivot_wider</span>(<span class="at">names_from =</span> location_property, <span class="at">values_from =</span> value)</span>
<span id="cb15-9"><a href="#cb15-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-10"><a href="#cb15-10" aria-hidden="true" tabindex="-1"></a>traits_with_locations <span class="ot">&lt;-</span> </span>
<span id="cb15-11"><a href="#cb15-11" aria-hidden="true" tabindex="-1"></a>  austraits<span class="sc">$</span>traits <span class="sc">%&gt;%</span> </span>
<span id="cb15-12"><a href="#cb15-12" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">left_join</span>(locations, <span class="at">by =</span> <span class="fu">c</span>(<span class="st">"dataset_id"</span>, <span class="st">"location_id"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="contexts" class="level3">
<h3 class="anchored" data-anchor-id="contexts">Contexts</h3>
<p>Joining the context table to the traits table is best done using the austraits.R function, as there are 5 identifiers in the traits table (<code>entity_context_id</code>, <code>method_context_id</code>, <code>temporal_context_id</code>, <code>plot_context_id</code> and <code>treatment_context_id</code>) that link to two columns in the contexts table (<code>link_vals</code> and <code>link_id</code>).</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb16"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a>traits_with_contexts <span class="ot">&lt;-</span> </span>
<span id="cb16-2"><a href="#cb16-2" aria-hidden="true" tabindex="-1"></a>  austraits <span class="sc">%&gt;%</span> austraits<span class="sc">::</span><span class="fu">join_contexts</span>()</span>
<span id="cb16-3"><a href="#cb16-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb16-4"><a href="#cb16-4" aria-hidden="true" tabindex="-1"></a>traits_with_contexts<span class="sc">$</span>traits <span class="sc">%&gt;%</span> </span>
<span id="cb16-5"><a href="#cb16-5" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">select</span>(<span class="sc">-</span>dplyr<span class="sc">::</span><span class="fu">all_of</span>(<span class="fu">c</span>(<span class="st">"temporal_context_id"</span>, <span class="st">"method_context_id"</span>, <span class="st">"treatment_context_id"</span>, <span class="st">"entity_context_id"</span>, <span class="st">"plot_context_id"</span>)))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>If you wish to include fewer context properties, you have to manually select the columns you wish to retain/lose.</p>
</section>
<section id="methods" class="level3">
<h3 class="anchored" data-anchor-id="methods">Methods</h3>
<p>To just add trait methods:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a>traits_with_methods <span class="ot">&lt;-</span> </span>
<span id="cb17-2"><a href="#cb17-2" aria-hidden="true" tabindex="-1"></a>  austraits <span class="sc">%&gt;%</span> austraits<span class="sc">::</span><span class="fu">join_methods</span>()</span>
<span id="cb17-3"><a href="#cb17-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-4"><a href="#cb17-4" aria-hidden="true" tabindex="-1"></a>traits_with_methods<span class="sc">$</span>traits</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>To add all information within the methods table:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb18"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a>traits_with_all_methods <span class="ot">&lt;-</span> </span>
<span id="cb18-2"><a href="#cb18-2" aria-hidden="true" tabindex="-1"></a>  austraits<span class="sc">$</span>traits <span class="sc">%&gt;%</span> </span>
<span id="cb18-3"><a href="#cb18-3" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">left_join</span>(austraits<span class="sc">$</span>methods, </span>
<span id="cb18-4"><a href="#cb18-4" aria-hidden="true" tabindex="-1"></a>            <span class="at">by =</span> <span class="fu">c</span>(<span class="st">"dataset_id"</span>, <span class="st">"trait_name"</span>, <span class="st">"method_id"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>As there are quite a few columns in the method table with lengthy text fields, it is worth first viewing the methods table and confirming you want information like <code>sampling_strategy</code> merged into the traits table.</p>
<p>“dataset_id”, “trait_name”, “methods”, “description”, “sampling_strategy”, “source_primary_key”, “source_primary_citation”, “source_secondary_key”, “source_secondary_citation”, “source_original_dataset_key”, “source_original_dataset_citation”, “data_collectors”, “assistants”, “austraits_curators”</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb19"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a>traits_with_all_methods <span class="ot">&lt;-</span> </span>
<span id="cb19-2"><a href="#cb19-2" aria-hidden="true" tabindex="-1"></a>  austraits<span class="sc">$</span>traits <span class="sc">%&gt;%</span> </span>
<span id="cb19-3"><a href="#cb19-3" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">left_join</span>(austraits<span class="sc">$</span>methods <span class="sc">%&gt;%</span> </span>
<span id="cb19-4"><a href="#cb19-4" aria-hidden="true" tabindex="-1"></a>              dplyr<span class="sc">::</span><span class="fu">select</span>(<span class="st">"dataset_id"</span>, <span class="st">"trait_name"</span>, <span class="st">"methods"</span>, <span class="st">"method_id"</span>, <span class="st">"description"</span>,  <span class="st">"source_primary_citation"</span>, <span class="st">"data_collectors"</span>),</span>
<span id="cb19-5"><a href="#cb19-5" aria-hidden="true" tabindex="-1"></a>            <span class="at">by =</span> <span class="fu">c</span>(<span class="st">"dataset_id"</span>, <span class="st">"trait_name"</span>, <span class="st">"method_id"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="taxa" class="level3">
<h3 class="anchored" data-anchor-id="taxa">Taxa</h3>
<p>To add family, genus, taxon_rank and establishment_means:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb20"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a>austraits_with_taxa <span class="ot">&lt;-</span></span>
<span id="cb20-2"><a href="#cb20-2" aria-hidden="true" tabindex="-1"></a>  austraits <span class="sc">%&gt;%</span> austraits<span class="sc">::</span><span class="fu">join_taxonomy</span>()</span>
<span id="cb20-3"><a href="#cb20-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb20-4"><a href="#cb20-4" aria-hidden="true" tabindex="-1"></a>austraits_with_taxa<span class="sc">$</span>traits</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>To add all columns from austraits$taxa:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb21"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a>austraits_with_taxa <span class="ot">&lt;-</span></span>
<span id="cb21-2"><a href="#cb21-2" aria-hidden="true" tabindex="-1"></a>  austraits<span class="sc">$</span>traits <span class="sc">%&gt;%</span></span>
<span id="cb21-3"><a href="#cb21-3" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">left_join</span>(austraits<span class="sc">$</span>taxa)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
</section>
<section id="plotting-data" class="level2" data-number="29.7">
<h2 data-number="29.7" class="anchored" data-anchor-id="plotting-data"><span class="header-section-number">29.7</span> Plotting data</h2>
<section id="plotting-locations" class="level3">
<h3 class="anchored" data-anchor-id="plotting-locations">Plotting locations</h3>
<p>To plot locations, you need to begin by merging on the latitude &amp; longitude data from austraits$locations.</p>
<section id="plot-locations-by-trait" class="level4">
<h4 class="anchored" data-anchor-id="plot-locations-by-trait">Plot locations by trait</h4>
<div class="cell">
<div class="sourceCode cell-code" id="cb22"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a>data <span class="ot">&lt;-</span></span>
<span id="cb22-2"><a href="#cb22-2" aria-hidden="true" tabindex="-1"></a>  austraits <span class="sc">%&gt;%</span> </span>
<span id="cb22-3"><a href="#cb22-3" aria-hidden="true" tabindex="-1"></a>     austraits<span class="sc">::</span><span class="fu">extract_trait</span>(<span class="at">trait_names =</span> <span class="fu">c</span>(<span class="st">"leaf_mass_per_area"</span>)) <span class="sc">%&gt;%</span></span>
<span id="cb22-4"><a href="#cb22-4" aria-hidden="true" tabindex="-1"></a>     austraits<span class="sc">::</span><span class="fu">join_locations</span>()</span>
<span id="cb22-5"><a href="#cb22-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-6"><a href="#cb22-6" aria-hidden="true" tabindex="-1"></a>data<span class="sc">$</span>traits <span class="sc">%&gt;%</span> austraits<span class="sc">::</span><span class="fu">plot_locations</span>(<span class="st">"trait_name"</span>)</span>
<span id="cb22-7"><a href="#cb22-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-8"><a href="#cb22-8" aria-hidden="true" tabindex="-1"></a><span class="co"># or</span></span>
<span id="cb22-9"><a href="#cb22-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-10"><a href="#cb22-10" aria-hidden="true" tabindex="-1"></a>data <span class="ot">&lt;-</span></span>
<span id="cb22-11"><a href="#cb22-11" aria-hidden="true" tabindex="-1"></a>  austraits <span class="sc">%&gt;%</span></span>
<span id="cb22-12"><a href="#cb22-12" aria-hidden="true" tabindex="-1"></a>     austraits<span class="sc">::</span><span class="fu">extract_trait</span>(<span class="at">trait_names =</span> <span class="fu">c</span>(<span class="st">"leaf_mass_per_area"</span>, <span class="st">"plant_height"</span>)) <span class="sc">%&gt;%</span></span>
<span id="cb22-13"><a href="#cb22-13" aria-hidden="true" tabindex="-1"></a>     austraits<span class="sc">::</span><span class="fu">join_locations</span>()</span>
<span id="cb22-14"><a href="#cb22-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-15"><a href="#cb22-15" aria-hidden="true" tabindex="-1"></a>data<span class="sc">$</span>traits <span class="sc">%&gt;%</span> austraits<span class="sc">::</span><span class="fu">plot_locations</span>(<span class="st">"trait_name"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="plot-locations-by-study" class="level4">
<h4 class="anchored" data-anchor-id="plot-locations-by-study">Plot locations by study</h4>
<div class="cell">
<div class="sourceCode cell-code" id="cb23"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a>data <span class="ot">&lt;-</span></span>
<span id="cb23-2"><a href="#cb23-2" aria-hidden="true" tabindex="-1"></a>  austraits <span class="sc">%&gt;%</span></span>
<span id="cb23-3"><a href="#cb23-3" aria-hidden="true" tabindex="-1"></a>     austraits<span class="sc">::</span><span class="fu">extract_dataset</span>(<span class="st">"Westoby_2014"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb23-4"><a href="#cb23-4" aria-hidden="true" tabindex="-1"></a>     austraits<span class="sc">::</span><span class="fu">join_locations</span>()</span>
<span id="cb23-5"><a href="#cb23-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-6"><a href="#cb23-6" aria-hidden="true" tabindex="-1"></a>data<span class="sc">$</span>traits <span class="sc">%&gt;%</span> austraits<span class="sc">::</span><span class="fu">plot_locations</span>(<span class="st">"dataset_id"</span>)</span>
<span id="cb23-7"><a href="#cb23-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-8"><a href="#cb23-8" aria-hidden="true" tabindex="-1"></a>data <span class="ot">&lt;-</span></span>
<span id="cb23-9"><a href="#cb23-9" aria-hidden="true" tabindex="-1"></a>  austraits <span class="sc">%&gt;%</span></span>
<span id="cb23-10"><a href="#cb23-10" aria-hidden="true" tabindex="-1"></a>     austraits<span class="sc">::</span><span class="fu">extract_dataset</span>(<span class="st">"Schulze_2014"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb23-11"><a href="#cb23-11" aria-hidden="true" tabindex="-1"></a>     austraits<span class="sc">::</span><span class="fu">join_locations</span>()</span>
<span id="cb23-12"><a href="#cb23-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-13"><a href="#cb23-13" aria-hidden="true" tabindex="-1"></a>data<span class="sc">$</span>traits <span class="sc">%&gt;%</span> austraits<span class="sc">::</span><span class="fu">plot_locations</span>(<span class="st">"dataset_id"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
</section>
<section id="an-example-looking-at-trait-climate-gradients" class="level3">
<h3 class="anchored" data-anchor-id="an-example-looking-at-trait-climate-gradients">An example looking at trait-climate gradients</h3>
<p>A simple workflow allows one to look at <a href="traits_and_climate_example.html">trait values across a climate gradient</a></p>
</section>
<section id="an-example-with-ala-data" class="level3">
<h3 class="anchored" data-anchor-id="an-example-with-ala-data">An example with ALA data</h3>
<p>A recent tutorial posted by ALA shows how one can combine AusTraits trait data and ALA spatial occurrence data:</p>
<p>https://labs.ala.org.au/posts/2023-08-28_alternatives-to-box-plots/post.html</p>
<p>We’ve adopted it <a href="spatial_data_example.html">here</a>.</p>
</section>
</section>
<section id="summarising-data" class="level2" data-number="29.8">
<h2 data-number="29.8" class="anchored" data-anchor-id="summarising-data"><span class="header-section-number">29.8</span> Summarising data</h2>
<section id="summarising-traits-by-taxon-counts-dataset_id-etc." class="level3">
<h3 class="anchored" data-anchor-id="summarising-traits-by-taxon-counts-dataset_id-etc.">Summarising traits by taxon counts, dataset_id, etc.</h3>
<p>As described above, you can summarise all of AusTraits by trait name:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb24"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a>austraits<span class="sc">::</span><span class="fu">summarise_austraits</span>(austraits, <span class="st">"trait_name"</span>) <span class="sc">%&gt;%</span> <span class="fu">slice</span>(<span class="dv">100</span><span class="sc">:</span><span class="dv">130</span>)</span>
<span id="cb24-2"><a href="#cb24-2" aria-hidden="true" tabindex="-1"></a>austraits<span class="sc">::</span><span class="fu">summarise_austraits</span>(austraits, <span class="at">var =</span>  <span class="st">"family"</span>) <span class="sc">%&gt;%</span> <span class="fu">slice</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">20</span>)</span>
<span id="cb24-3"><a href="#cb24-3" aria-hidden="true" tabindex="-1"></a>austraits<span class="sc">::</span><span class="fu">summarise_austraits</span>(austraits, <span class="at">var =</span>  <span class="st">"genus"</span>) <span class="sc">%&gt;%</span> <span class="fu">slice</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">20</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Or just look at info for nitrogen-related traits:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb25"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a>austraits<span class="sc">::</span><span class="fu">lookup_trait</span>(austraits, <span class="st">"_N_"</span>) <span class="ot">-&gt;</span> N_traits</span>
<span id="cb25-2"><a href="#cb25-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-3"><a href="#cb25-3" aria-hidden="true" tabindex="-1"></a>austraits<span class="sc">$</span>traits <span class="sc">%&gt;%</span> </span>
<span id="cb25-4"><a href="#cb25-4" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">filter</span>(trait_name <span class="sc">%in%</span> N_traits) <span class="sc">%&gt;%</span> </span>
<span id="cb25-5"><a href="#cb25-5" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">distinct</span>(taxon_name, trait_name) <span class="sc">%&gt;%</span> </span>
<span id="cb25-6"><a href="#cb25-6" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">mutate</span>(<span class="at">counter =</span> <span class="dv">1</span>) <span class="sc">%&gt;%</span> </span>
<span id="cb25-7"><a href="#cb25-7" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">group_by</span>(trait_name) <span class="sc">%&gt;%</span> </span>
<span id="cb25-8"><a href="#cb25-8" aria-hidden="true" tabindex="-1"></a>    dplyr<span class="sc">::</span><span class="fu">mutate</span>(<span class="at">counter =</span> <span class="fu">sum</span>(counter)) <span class="sc">%&gt;%</span> </span>
<span id="cb25-9"><a href="#cb25-9" aria-hidden="true" tabindex="-1"></a>    dplyr<span class="sc">::</span><span class="fu">distinct</span>(trait_name, counter) <span class="sc">%&gt;%</span> </span>
<span id="cb25-10"><a href="#cb25-10" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">ungroup</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Or look at how many datasets have each trait:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb26"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb26-1"><a href="#cb26-1" aria-hidden="true" tabindex="-1"></a>austraits<span class="sc">::</span><span class="fu">lookup_trait</span>(austraits, <span class="st">"_N_"</span>) <span class="ot">-&gt;</span> N_traits</span>
<span id="cb26-2"><a href="#cb26-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-3"><a href="#cb26-3" aria-hidden="true" tabindex="-1"></a>austraits<span class="sc">$</span>traits <span class="sc">%&gt;%</span> </span>
<span id="cb26-4"><a href="#cb26-4" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">filter</span>(trait_name <span class="sc">%in%</span> N_traits) <span class="sc">%&gt;%</span> </span>
<span id="cb26-5"><a href="#cb26-5" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">distinct</span>(dataset_id, trait_name) <span class="sc">%&gt;%</span> </span>
<span id="cb26-6"><a href="#cb26-6" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">mutate</span>(<span class="at">counter =</span> <span class="dv">1</span>) <span class="sc">%&gt;%</span> </span>
<span id="cb26-7"><a href="#cb26-7" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">group_by</span>(trait_name) <span class="sc">%&gt;%</span> </span>
<span id="cb26-8"><a href="#cb26-8" aria-hidden="true" tabindex="-1"></a>    dplyr<span class="sc">::</span><span class="fu">mutate</span>(<span class="at">counter =</span> <span class="fu">sum</span>(counter)) <span class="sc">%&gt;%</span> </span>
<span id="cb26-9"><a href="#cb26-9" aria-hidden="true" tabindex="-1"></a>    dplyr<span class="sc">::</span><span class="fu">distinct</span>(trait_name, counter) <span class="sc">%&gt;%</span> </span>
<span id="cb26-10"><a href="#cb26-10" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">ungroup</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>If you’ve sourced the file <code>extra_functions.R</code>, there are a few functions that allow you to summarise trait values.</p>
<p>For instance, <code>categorical_summary</code> indicates how many times a specific trait value is reported for a given taxa (across all datasets):</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb27"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true" tabindex="-1"></a>cat_summary <span class="ot">&lt;-</span> <span class="fu">categorical_summary</span>(austraits, <span class="st">"resprouting_capacity"</span>)</span>
<span id="cb27-2"><a href="#cb27-2" aria-hidden="true" tabindex="-1"></a><span class="fu">View</span>(cat_summary)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Alternatively, create a wider matrix with possible trait values as columns:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb28"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb28-1"><a href="#cb28-1" aria-hidden="true" tabindex="-1"></a>categorical_summary_wider <span class="ot">&lt;-</span> </span>
<span id="cb28-2"><a href="#cb28-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">categorical_summary_by_value</span>(austraits, <span class="st">"resprouting_capacity"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb28-3"><a href="#cb28-3" aria-hidden="true" tabindex="-1"></a>    tidyr<span class="sc">::</span><span class="fu">pivot_wider</span>(<span class="at">names_from =</span> value_tmp, <span class="at">values_from =</span> replicates)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We’re hesitant to write functions to summarise numeric traits, as this quickly requires statistical assumptions.</p>
<p>The datasets that comprise AusTraits were collected by different people, with different number of replicates, different entity types are reported - one dataset might include 20 measurements on individuals for a trait and another dataset might have submitted a single population-level mean derived from 5 measurements.</p>
<p>How do you take the mean of these trait values?</p>
<p>Do you want to include both data from experiments and plants growing under natural conditions? This information is recorded in the <code>basis_of_record</code> column.</p>
<p>One algorithm we’ve developed is:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb29"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb29-1"><a href="#cb29-1" aria-hidden="true" tabindex="-1"></a>weighted <span class="ot">&lt;-</span> <span class="fu">austraits_weighted_means</span>(austraits, <span class="fu">c</span>(<span class="st">"leaf_mass_per_area"</span>, <span class="st">"leaf_length"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>This function first groups values at the site level, then at the taxon level. For trait data sourced from floras, where trait values are documented as a minimum and maximum value, it simply takes the mean of these as an additional mean.</p>
</section>
</section>
<section id="a-complexity-pivoting-datasets" class="level2" data-number="29.9">
<h2 data-number="29.9" class="anchored" data-anchor-id="a-complexity-pivoting-datasets"><span class="header-section-number">29.9</span> A complexity: pivoting datasets</h2>
<p>AusTraits datasets can be pivoted from <code>long</code> to <code>wide</code> format:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb30"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb30-1"><a href="#cb30-1" aria-hidden="true" tabindex="-1"></a>Farrell_2017_values <span class="ot">&lt;-</span></span>
<span id="cb30-2"><a href="#cb30-2" aria-hidden="true" tabindex="-1"></a>  austraits <span class="sc">%&gt;%</span></span>
<span id="cb30-3"><a href="#cb30-3" aria-hidden="true" tabindex="-1"></a>     austraits<span class="sc">::</span><span class="fu">extract_dataset</span>(<span class="st">"Farrell_2017"</span>)</span>
<span id="cb30-4"><a href="#cb30-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-5"><a href="#cb30-5" aria-hidden="true" tabindex="-1"></a>Farrell_2017_pivoted <span class="ot">&lt;-</span> </span>
<span id="cb30-6"><a href="#cb30-6" aria-hidden="true" tabindex="-1"></a>  Farrell_2017_values<span class="sc">$</span>traits <span class="sc">%&gt;%</span></span>
<span id="cb30-7"><a href="#cb30-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">trait_pivot_wider</span>()</span>
<span id="cb30-8"><a href="#cb30-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-9"><a href="#cb30-9" aria-hidden="true" tabindex="-1"></a><span class="fu">View</span>(Farrell_2017_pivoted)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>This example pivots “nicely” as all observations have <code>entity_type = individual</code>.</p>
<p>Compare this first example to the dataset <code>Edwards_2000</code> which includes individual-level, population-level, and species-level observations:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb31"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb31-1"><a href="#cb31-1" aria-hidden="true" tabindex="-1"></a>Edwards_2000_values <span class="ot">&lt;-</span></span>
<span id="cb31-2"><a href="#cb31-2" aria-hidden="true" tabindex="-1"></a>  austraits <span class="sc">%&gt;%</span></span>
<span id="cb31-3"><a href="#cb31-3" aria-hidden="true" tabindex="-1"></a>     austraits<span class="sc">::</span><span class="fu">extract_dataset</span>(<span class="st">"Edwards_2000"</span>)</span>
<span id="cb31-4"><a href="#cb31-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-5"><a href="#cb31-5" aria-hidden="true" tabindex="-1"></a>Edwards_2000_pivoted <span class="ot">&lt;-</span> </span>
<span id="cb31-6"><a href="#cb31-6" aria-hidden="true" tabindex="-1"></a>  Edwards_2000_values<span class="sc">$</span>traits <span class="sc">%&gt;%</span></span>
<span id="cb31-7"><a href="#cb31-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">trait_pivot_wider</span>()</span>
<span id="cb31-8"><a href="#cb31-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-9"><a href="#cb31-9" aria-hidden="true" tabindex="-1"></a><span class="fu">View</span>(Edwards_2000_pivoted)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The values at the individual, population and species level do not collapse together, because traits measured on different <code>entity_types</code> have separate <code>observation_id</code>’s.</p>
<p>One of the core identifiers assigned to data points is the <code>observation_id</code>. An observation is a collection of measurements made on a specific entity at a single point in time.</p>
<p>Observation_id’s are therefore unique combinations of:</p>
<ul>
<li>dataset_id</li>
<li>source_id</li>
<li>entity_type</li>
<li>taxon_name</li>
<li>population_id (location_id, plot_context_id, treatment_context_id)</li>
<li>individual_id</li>
<li>basis_of_record</li>
<li>entity_context_id</li>
<li>life_stage</li>
<li>temporal_context_id</li>
<li>collection_date</li>
<li>original_name</li>
</ul>
<p>If a single dataset includes traits that are attributed to different entity types, they are assigned separate <code>observation_id</code>’s. For instance, many datasets are comprised of individual-level physiological trait data and a column <code>growth_form</code>, documenting the growth form (i.e.&nbsp;tree, shrub, herb, etc.) of each <em>species</em>.</p>
<p>We’re developing a function that collapses the pivoted data into a more condensed table, but this loses some of the metadata:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb32"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb32-1"><a href="#cb32-1" aria-hidden="true" tabindex="-1"></a>Edwards_2000_pivoted_merged <span class="ot">&lt;-</span></span>
<span id="cb32-2"><a href="#cb32-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">merge_entity_types</span>(<span class="st">"Edwards_2000"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<ul>
<li><p>This function will duplicate any “higher-entity” trait values (e.g.&nbsp;A single species-level value is filled in for all individuals or populations)</p></li>
<li><p>Metadata fields like <code>entity_type</code>, <code>value_type</code> is only retained if it is identical for all measurements</p></li>
</ul>
<div class="cell">
<div class="sourceCode cell-code" id="cb33"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb33-1"><a href="#cb33-1" aria-hidden="true" tabindex="-1"></a>Westoby_2014_pivoted_merged <span class="ot">&lt;-</span></span>
<span id="cb33-2"><a href="#cb33-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">merge_entity_types</span>(<span class="st">"Westoby_2014"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="intepreting-trait-names-taxon-names" class="level2" data-number="29.10">
<h2 data-number="29.10" class="anchored" data-anchor-id="intepreting-trait-names-taxon-names"><span class="header-section-number">29.10</span> Intepreting trait names, taxon names</h2>
<section id="trait-dictionary" class="level3">
<h3 class="anchored" data-anchor-id="trait-dictionary">Trait dictionary</h3>
<p>The AusTraits pipeline requires a trait dictionary that documents 4 pieces of information about each trait:</p>
<ul>
<li>trait name&nbsp;</li>
<li>trait type (categorical vs numeric)&nbsp;</li>
<li>allowable trait values (for categorical traits)&nbsp;</li>
<li>allowable trait range and units&nbsp;</li>
</ul>
<p>The trait dictionary embedded within AusTraits also has:</p>
<ul>
<li>trait labels&nbsp;</li>
<li>trait definitions&nbsp;</li>
<li>definitions for all categorical trait values&nbsp;</li>
</ul>
<p>Together these clarify each “trait concept”, which we define as: “a circumscribed set of trait measurements”. Much like a taxon concept, delimits a collection of organisms, a trait concept delimits a collection of trait values pertaining to a distinct characteristic of a specific part of an organism (cell, tissue, organ, or whole organism)”</p>
<p>The <a href="http:///w3id.org/APD">AusTraits Plant Dictionary (APD)</a> is a formal vocabulary for the AusTraits trait definitions.</p>
<p>It also includes:&nbsp;</p>
<ul>
<li>keywords&nbsp;</li>
<li>plant structure measured&nbsp;</li>
<li>characteristic measured&nbsp;</li>
<li>references&nbsp;</li>
<li>links to the same (or similar) trait concepts in other databases and dictionaries&nbsp;</li>
</ul>
</section>
<section id="understanding-taxon-names" class="level3">
<h3 class="anchored" data-anchor-id="understanding-taxon-names">Understanding taxon names</h3>
<p>AusTraits uses the taxon names in the Australian Plant Census (APC) and the Australian Plant Names Index (APNI).</p>
<p>The R package <a href="https://github.com/traitecoevo/APCalign">APCalign</a> is used to align and update taxon names submitted to AusTraits with those in the APC/APNI.</p>
<p>APC align can also now be installed directly from CRAN</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb34"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb34-1"><a href="#cb34-1" aria-hidden="true" tabindex="-1"></a><span class="fu">install.packages</span>(<span class="st">"APCalign"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>There are two key components to the workflow:&nbsp;</p>
<ol type="1">
<li>aligning names&nbsp;</li>
</ol>
<ul>
<li><p>syntax is standardised, including for phrase names&nbsp;</p></li>
<li><p>most spelling mistakes are corrected&nbsp;</p></li>
<li><p>names that indicate the plant can only be identified to genus are reformatted to <code>genus sp. [available notes; dataset_id]</code>&nbsp;</p>
<ol type="1">
<li><p>they are linked to an APC-accepted genus but not to an APC-accepted binomial.&nbsp;</p></li>
<li><p>they include the dataset_id so people don’t mistakenly group all <code>Eucalyptus sp.</code> as a single “species”&nbsp;</p></li>
</ol></li>
</ul>
<ol start="2" type="1">
<li>updating names&nbsp;</li>
</ol>
<ul>
<li>all aligned names that are in the APC, but that have a <code>taxonomic status</code> other than <code>accepted</code> are updated to their currently accepted name.&nbsp;</li>
</ul>
<p>Examples:&nbsp;</p>
<p>Identical <code>genus sp.</code> inputs from disparate datasets are given unique “names”:&nbsp;</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb35"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb35-1"><a href="#cb35-1" aria-hidden="true" tabindex="-1"></a>austraits<span class="sc">$</span>traits <span class="sc">%&gt;%</span></span>
<span id="cb35-2"><a href="#cb35-2" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">filter</span>(stringr<span class="sc">::</span><span class="fu">str_detect</span>(original_name, <span class="st">"Eucalyptus sp</span><span class="sc">\\</span><span class="st">.$"</span>)) <span class="sc">%&gt;%</span></span>
<span id="cb35-3"><a href="#cb35-3" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">distinct</span>(dataset_id, taxon_name, original_name) <span class="sc">%&gt;%</span></span>
<span id="cb35-4"><a href="#cb35-4" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">filter</span>(original_name <span class="sc">!=</span> taxon_name) </span>
<span id="cb35-5"><a href="#cb35-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb35-6"><a href="#cb35-6" aria-hidden="true" tabindex="-1"></a><span class="co"># A tibble: 4 × 3</span></span>
<span id="cb35-7"><a href="#cb35-7" aria-hidden="true" tabindex="-1"></a>  dataset_id     taxon_name                      original_name </span>
<span id="cb35-8"><a href="#cb35-8" aria-hidden="true" tabindex="-1"></a>  <span class="sc">&lt;</span>chr<span class="sc">&gt;</span>          <span class="er">&lt;</span>chr<span class="sc">&gt;</span>                           <span class="er">&lt;</span>chr<span class="sc">&gt;</span>         </span>
<span id="cb35-9"><a href="#cb35-9" aria-hidden="true" tabindex="-1"></a><span class="dv">1</span> Clarke_2015    Eucalyptus sp. [Clarke_2015]    Eucalyptus sp.</span>
<span id="cb35-10"><a href="#cb35-10" aria-hidden="true" tabindex="-1"></a><span class="dv">2</span> Nolan_2022     Eucalyptus sp. [Nolan_2022]     Eucalyptus sp.</span>
<span id="cb35-11"><a href="#cb35-11" aria-hidden="true" tabindex="-1"></a><span class="dv">3</span> Soper_2014     Eucalyptus sp. [Soper_2014]     Eucalyptus sp.</span>
<span id="cb35-12"><a href="#cb35-12" aria-hidden="true" tabindex="-1"></a><span class="dv">4</span> Zieminska_2015 Eucalyptus sp. [Zieminska_2015] Eucalyptus sp.</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Outdated names are updated:&nbsp;</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb36"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb36-1"><a href="#cb36-1" aria-hidden="true" tabindex="-1"></a>austraits<span class="sc">$</span>traits <span class="sc">%&gt;%</span></span>
<span id="cb36-2"><a href="#cb36-2" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">filter</span>(stringr<span class="sc">::</span><span class="fu">str_detect</span>(original_name, <span class="st">"Dryandra"</span>)) <span class="sc">%&gt;%</span></span>
<span id="cb36-3"><a href="#cb36-3" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">distinct</span>(taxon_name, original_name) <span class="sc">%&gt;%</span></span>
<span id="cb36-4"><a href="#cb36-4" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">filter</span>(original_name <span class="sc">!=</span> taxon_name) <span class="sc">%&gt;%</span> <span class="fu">slice</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">15</span>)</span>
<span id="cb36-5"><a href="#cb36-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb36-6"><a href="#cb36-6" aria-hidden="true" tabindex="-1"></a><span class="co"># A tibble: 152 × 2</span></span>
<span id="cb36-7"><a href="#cb36-7" aria-hidden="true" tabindex="-1"></a>   taxon_name                   original_name                </span>
<span id="cb36-8"><a href="#cb36-8" aria-hidden="true" tabindex="-1"></a>   <span class="sc">&lt;</span>chr<span class="sc">&gt;</span>                        <span class="er">&lt;</span>chr<span class="sc">&gt;</span>                        </span>
<span id="cb36-9"><a href="#cb36-9" aria-hidden="true" tabindex="-1"></a> <span class="dv">1</span> Banksia acanthopoda          Dryandra acanthopoda         </span>
<span id="cb36-10"><a href="#cb36-10" aria-hidden="true" tabindex="-1"></a> <span class="dv">2</span> Banksia anatona              Dryandra anatona             </span>
<span id="cb36-11"><a href="#cb36-11" aria-hidden="true" tabindex="-1"></a> <span class="dv">3</span> Banksia arborea              Dryandra arborea             </span>
<span id="cb36-12"><a href="#cb36-12" aria-hidden="true" tabindex="-1"></a> <span class="dv">4</span> Banksia brunnea              Dryandra arctotidis          </span>
<span id="cb36-13"><a href="#cb36-13" aria-hidden="true" tabindex="-1"></a> <span class="dv">5</span> Banksia armata               Dryandra armata              </span>
<span id="cb36-14"><a href="#cb36-14" aria-hidden="true" tabindex="-1"></a> <span class="dv">6</span> Banksia armata var. armata   Dryandra armata var. armata  </span>
<span id="cb36-15"><a href="#cb36-15" aria-hidden="true" tabindex="-1"></a> <span class="dv">7</span> Banksia armata var. ignicida Dryandra armata var. ignicida</span>
<span id="cb36-16"><a href="#cb36-16" aria-hidden="true" tabindex="-1"></a> <span class="dv">8</span> Banksia aurantia             Dryandra aurantia            </span>
<span id="cb36-17"><a href="#cb36-17" aria-hidden="true" tabindex="-1"></a> <span class="dv">9</span> Banksia biterax              Dryandra baxteri             </span>
<span id="cb36-18"><a href="#cb36-18" aria-hidden="true" tabindex="-1"></a><span class="dv">10</span> Banksia bipinnatifida        Dryandra bipinnatifida </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Phrase name syntax across datasets is aligned:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb37"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb37-1"><a href="#cb37-1" aria-hidden="true" tabindex="-1"></a>austraits<span class="sc">$</span>traits <span class="sc">%&gt;%</span></span>
<span id="cb37-2"><a href="#cb37-2" aria-hidden="true" tabindex="-1"></a>     dplyr<span class="sc">::</span><span class="fu">filter</span>(stringr<span class="sc">::</span><span class="fu">str_detect</span>(taxon_name, <span class="st">"Argyrodendron sp. Whyanbeel"</span>)) <span class="sc">%&gt;%</span></span>
<span id="cb37-3"><a href="#cb37-3" aria-hidden="true" tabindex="-1"></a>     dplyr<span class="sc">::</span><span class="fu">distinct</span>(taxon_name, original_name)</span>
<span id="cb37-4"><a href="#cb37-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb37-5"><a href="#cb37-5" aria-hidden="true" tabindex="-1"></a><span class="co"># A tibble: 6 × 2</span></span>
<span id="cb37-6"><a href="#cb37-6" aria-hidden="true" tabindex="-1"></a>  taxon_name                                       original_name                                    </span>
<span id="cb37-7"><a href="#cb37-7" aria-hidden="true" tabindex="-1"></a>  <span class="sc">&lt;</span>chr<span class="sc">&gt;</span>                                            <span class="er">&lt;</span>chr<span class="sc">&gt;</span>                                            </span>
<span id="cb37-8"><a href="#cb37-8" aria-hidden="true" tabindex="-1"></a><span class="dv">1</span> Argyrodendron sp. <span class="fu">Whyanbeel</span> (B.P.Hyland RFK1106) <span class="fu">Argyrodendron</span> (Whyanbeel)                        </span>
<span id="cb37-9"><a href="#cb37-9" aria-hidden="true" tabindex="-1"></a><span class="dv">2</span> Argyrodendron sp. <span class="fu">Whyanbeel</span> (B.P.Hyland RFK1106) Argyrodendron <span class="fu">ssp.</span> (Whyanbeel BH 1106RFK)        </span>
<span id="cb37-10"><a href="#cb37-10" aria-hidden="true" tabindex="-1"></a><span class="dv">3</span> Argyrodendron sp. <span class="fu">Whyanbeel</span> (B.P.Hyland RFK1106) Argyrodendron Whyanbeel                          </span>
<span id="cb37-11"><a href="#cb37-11" aria-hidden="true" tabindex="-1"></a><span class="dv">4</span> Argyrodendron sp. <span class="fu">Whyanbeel</span> (B.P.Hyland RFK1106) Argyrodendron <span class="fu">sp.</span> (Whyanbeel BH 1106RFK)         </span>
<span id="cb37-12"><a href="#cb37-12" aria-hidden="true" tabindex="-1"></a><span class="dv">5</span> Argyrodendron sp. <span class="fu">Whyanbeel</span> (B.P.Hyland RFK1106) Argyrodendron sp. <span class="fu">Whyanbeel</span> (B.P.Hyland RFK <span class="dv">1106</span>)</span>
<span id="cb37-13"><a href="#cb37-13" aria-hidden="true" tabindex="-1"></a><span class="dv">6</span> Argyrodendron sp. <span class="fu">Whyanbeel</span> (B.P.Hyland RFK1106) Argyrodendron sp. <span class="fu">Whyanbeel</span> (B.P.Hyland RFK1106) </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>


</section>
</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
<nav class="page-navigation">
  <div class="nav-page nav-page-previous">
      <a href="./austraits_package.html" class="pagination-link">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text"><span class="chapter-number">28</span>&nbsp; <span class="chapter-title">The <code>austraits</code> package</span></span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
      <a href="./traits_and_climate_example.html" class="pagination-link">
        <span class="nav-page-text"><span class="chapter-number">30</span>&nbsp; <span class="chapter-title">Combining trait values and climate data</span></span> <i class="bi bi-arrow-right-short"></i>
      </a>
  </div>
</nav>
</div> <!-- /content -->
<footer class="footer">
  <div class="nav-footer">
    <div class="nav-footer-left">
      &nbsp;
    </div>   
    <div class="nav-footer-center">How to get help: <a href="https://traitecoevo.github.io/traits.build-book/help.html" class="uri">https://traitecoevo.github.io/traits.build-book/help.html</a><br>Copyright 2023, Daniel Falster and Elizabeth Wenk</div>
    <div class="nav-footer-right">
      &nbsp;
    </div>
  </div>
</footer>



</body></html>